libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/ActivityChecker.cs

```diff
@@ -0,0 +1,372 @@
+            }
+
+            return type?.ToLower() ?? string.Empty;
+        }
+
+        private static List<Diagnostic> CheckAttachments(JToken value)
+        {
+            var result = new List<Diagnostic>();
+
+            var attachmentsJsonList = NormalizedToList(value);
+
+            foreach (var attachmentsJson in attachmentsJsonList)
+            {
+                if (attachmentsJson is JObject attachmentsJsonJObj)
+                {
+                    result.AddRange(CheckAttachment(attachmentsJsonJObj));
+                }
+            }
+
+            return result;
+        }
+
+        private static List<Diagnostic> CheckAttachment(JObject lgJObj)
+        {
+            var result = new List<Diagnostic>();
+
+            var type = GetStructureType(lgJObj);
+
+            if (ActivityFactory.GenericCardTypeMapping.ContainsKey(type))
+            {
+                result.AddRange(CheckCardAtttachment(ActivityFactory.GenericCardTypeMapping[type], lgJObj));
+            }
+            else if (type == nameof(AdaptiveCard).ToLower())
+            {
+                try
+                {
+                    var parseResult = AdaptiveCard.FromJson(lgJObj.ToString());
+                    if (parseResult.Warnings.Count > 0)
+                    {
+                        foreach (var warning in parseResult.Warnings)
+                        {
+                            result.Add(BuildDiagnostic(warning.Message, false));
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    result.Add(BuildDiagnostic(e.Message));
+                }
+            }
+            else
+            {
+                result.Add(BuildDiagnostic($"'{type}' is not an attachment type.", false));
+            }
+
+            return result;
+        }
+
+        private static List<Diagnostic> CheckCardAtttachment(string type, JObject lgJObj)
+        {
+            var result = new List<Diagnostic>();
+
+            foreach (var item in lgJObj)
+            {
+                var property = item.Key.Trim().ToLower();
+                var value = item.Value;
+
+                switch (property)
+                {
+                    case "$type":
+                    case "title":
+                    case "subtitle":
+                    case "text":
+                    case "aspect":
+                    case "value":
+                    case "connectionname":
+                    case "image":
+                    case "images":
+                    case "media":
+                        break;
+                    case "buttons":
+                        result.AddRange(CheckButtons(value));
+                        break;
+
+                    case "autostart":
+                    case "shareable":
+                    case "autoloop":
+                        if (value.ToString().ToLower() != "true" && value.ToString().ToLower() != "false")
```

Comment:
Consider adding methods like "IsValidBooleanValue()" to make this code more readable.